---
title: "_systemPipeRdata_: Workflow templates and sample data"
author: "Author: Le Zhang, Daniela Cassol, and Thomas Girke"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output:
  BiocStyle::html_document:
    toc_float: true
    code_folding: show
    toc_depth: 4
package: systemPipeRdata
vignette: |
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{systemPipeRdata: Workflow templates and sample data}
  %\VignetteEngine{knitr::rmarkdown}
fontsize: 14pt
bibliography: bibtex.bib
---

```{css, echo=FALSE}
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
```

<!---
- Compile from command-line
Rscript -e "rmarkdown::render('systemPipeRdata.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('systemPipeRdata.Rmd', tangle=TRUE)"

-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=60, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")), 
    tidy.opts=list(width.cutoff=60), tidy=TRUE)
```

```{r setup, echo=FALSE, messages=FALSE, warnings=FALSE}
suppressPackageStartupMessages({
    library(systemPipeRdata)
})
```

# Introduction

[`systemPipeRdata`](https://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html)
provides data analysis workflow templates compatible with the
[`systemPipeR`](https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html)
software package [@H_Backman2016-bt]. The latter is a Workflow Management
System (WMS) for designing and running end-to-end analysis workflows with
automated report generation for a wide range of data analysis applications.
These workflow templates come equipped with sample data and the necessary
parameter files required to run a selected workflow. This setup simplifies the
learning process of using `systemPipeR`, facilitates testing of workflows, and
serves as a foundation for designing new workflows. The standardized directory
structure (Figure 1) utilized by the workflow templates and their sample data is outlined
in the ['Directory Structure'](https://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#3_Directory_structure) 
section of `systemPipeR's` main vignette.

<center><img src="results/directory.png"></center>

__Figure 1:__ Directory structure of`systemPipeR's` workflows. 

# Getting started

## Installation

The _`systemPipeRdata`_ package is available at [Bioconductor](http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html) and can be installed from within R as follows: 

```{r install, eval=FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("systemPipeRdata") 
```

## Loading package and documentation

```{r load_systemPipeRdata, eval=TRUE, messages=FALSE, warnings=FALSE}
library("systemPipeRdata") # Loads the package
```

```{r documentation_systemPipeRdata, eval=FALSE}
library(help="systemPipeRdata") # Lists package info
vignette("systemPipeRdata") # Opens vignette
```

# List workflow templates
A list of workflow templates included in `systemPipeRdata` can be generated as follows.

```{r check_availableWF, eval=TRUE}
availableWF()
```

# Use workflow templates

## Load a workflow
The chosen example below uses the `genWorkenvir` function from the
`systemPipeRdata` package to create an RNA-Seq workflow environment (selected
under `workflow="rnaseq"`) that is fully populated with a small test data set,
including FASTQ files, reference genome and annotation data. The name of the
resulting workflow directory can be specified under the `mydirname` argument.
The default `NULL` uses the name of the chosen workflow. An error is issued if
a directory of the same name and path exists already. After this, the userâ€™s R
session needs to be directed into the resulting rnaseq directory (here with
setwd).

```{r generate_workenvir, eval=FALSE}
library(systemPipeRdata)
genWorkenvir(workflow="rnaseq", mydirname="rnaseq")
setwd("rnaseq")
```

On Linux and OS X systems the same can be achieved from the command-line of a
terminal with the following commands.

```{bash generate_workenvir_from_shell, eval=FALSE}
$ Rscript -e "systemPipeRdata::genWorkenvir(workflow='rnaseq', mydirname='rnaseq')"
$ cd rnaseq
```

## Run and visualize workflow 

For running and working with `systemPipeR` workflows, users want to visit
[systemPipeR's main
vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html).
The following gives only a very brief preview on how to run workflows and create reports. 

After a workflow environment (directory) has been created and the corresponding
R session directed into the resulting directory (here `rnaseq`), the workflow
can be loaded from the included R Markdown file (here `systemPipeRNAseq.Rmd`).
This template provides common data analysis steps that are typical for RNA-Seq
workflows. Users have the options to add, remove, modify workflow steps by
applying those change to the `sal` workflow management container. 

```{r project_rnaseq, eval=FALSE}
library(systemPipeR)
sal <- SPRproject() 
sal <- importWF(sal, file_path = "systemPipeRNAseq.Rmd", verbose = FALSE)
```

Once the workflow has been loaded into `sal`, it can be executed from start to finish
(or partially) with the `runWF` command.

```{r run_rnaseq, eval=FALSE}
sal <- runWF(sal)
```
Workflows can be visualized as topology graphs using the `plotWF` function.

```{r plot_rnaseq, eval=FALSE}
plotWF(sal)
```

Scientific and technical reports can be generated with the `renderReport` and 
`renderLogs` functions respectively. Scientific reports can also be generated 
with the `render` function of the `rmarkdown` package. The technical reports are
based on log informatation that `systemPipeR` collects during workflow runs.

```{r report_rnaseq, eval=FALSE}
# Scietific report
sal <- renderReport(sal)
rmarkdown::render("systemPipeRNAseq.Rmd", clean = TRUE, output_format = "BiocStyle::html_document") 

# Technical (log) report
sal <- renderLogs(sal)
```

# Experimental workflow collection

A collection of experimental workflow templates is available on
[systemPipeR's GitHub Organization](https://github.com/systemPipeR). From
within R one can list this collection with the same `availableWF` function
used before.

```{r eval=FALSE, tidy=FALSE}
availableWF(github = TRUE)
```
<pre>
Experimental Workflow Templates in systemPipeR GitHub Organization:
       Workflow                                     Download URL
1     SPatacseq    https://github.com/systemPipeR/SPatacseq.git
2     SPclipseq    https://github.com/systemPipeR/SPclipseq.git
3      SPdenovo    https://github.com/systemPipeR/SPdenovo.git
4         SPhic    https://github.com/systemPipeR/SPhic.git
5   SPmetatrans    https://github.com/systemPipeR/SPmetatrans.git
6   SPmethylseq    https://github.com/systemPipeR/SPmethylseq.git
7    SPmirnaseq    https://github.com/systemPipeR/SPmirnaseq.git
8 SPpolyriboseq    https://github.com/systemPipeR/SPpolyriboseq.git
9    SPscrnaseq    https://github.com/systemPipeR/SPscrnaseq.git
</pre>

To download these packages, users can either use the below GitHub commands, 
or visit the corresponding GitHub page of a chosen workflow via the provided URLs,
download is as Zip file and uncompress it. 

```{bash eval=FALSE}
git clone <Download URL> # Provide under <...> URL of chosen workflow from table above.
cd <Workflow Name>
```

Similarly, then use `systemPipeR::SPRproject()` to start a new project.

# Useful functionalities

## Create workflow templates interactively

It is possible to create a new workflow environment from RStudio. This can be 
done by selecting `File -> New File -> R Markdown -> From Template -> systemPipeR New WorkFlow`.
This option creates a template workflow that has the expected directory structure of the
other templates (see [here](https://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#3_Directory_structure). 

![](results/rstudio.png)
__Figure 2:__ Selecting workflow template within RStudio.


## Return paths to sample data

The paths to the sample data provided by the `systemPipeRdata` package can be returned with the
the `pathList` function.

```{r return_samplepaths, eval=TRUE}
pathList()[1:2]
```

# Version information

```{r sessionInfo}
sessionInfo()
```

# Funding

This project was supported by funds from the National Institutes of Health (NIH) and the National Science Foundation (NSF).

# References
